src/protocol/aggregation/mod.rs

```diff
@@ -52,139 +64,241 @@ where
-            acc += &row.to_additive_sharing_in_large_field();
+    let equality_check_ctx = ctx
+        .narrow(&Step::ComputeEqualityChecks)
+        .set_total_records(num_records);
+
+    // Generate N streams for each bucket specified by the breakdown key (N = |breakdown_keys|).
+    // A stream is pipeline of contribution values multiplied by the "equality bit". An equality
+    // bit is a bit that is a share of 1 if the breakdown key matches the bucket, or 0 otherwise.
+    let streams = seq_join(
+        ctx.active_work(),
+        stream_iter(breakdown_keys)
+            .zip(stream_iter(contribution_values))
+            .enumerate()
+            .map(|(i, (bk, v))| {
+                let eq_ctx = &equality_check_ctx;
+                let c = ctx.clone();
+                async move {
+                    let equality_checks = check_everything(eq_ctx.clone(), i, &bk).await?;
+                    equality_bits_times_value(&c, equality_checks, num_buckets, v, num_records, i)
+                        .await
+                }
+            }),
+    );
+    // for each bucket stream, sum up the contribution values
+    streams
+        .try_fold(vec![S::ZERO; num_buckets], |mut acc, bucket| async move {
+            for (i, b) in bucket.into_iter().enumerate() {
+                acc[i] += &b;
+            }
             Ok(acc)
         })
-        .await?;
-
-    Ok(aggregate)
+        .await
 }
 
-fn get_gf2_value_bits<CV, BK>(
-    input_rows: &[SparseAggregateInputRow<CV, BK>],
-) -> Vec<BitDecomposed<Replicated<Gf2>>>
+async fn equality_bits_times_value<F, C, S>(
+    ctx: &C,
+    check_bits: BitDecomposed<S>,
+    num_buckets: usize,
+    value_bits: BitDecomposed<S>,
+    num_records: usize,
+    record_id: usize,
+) -> Result<Vec<S>, Error>
 where
-    CV: GaloisField,
-    BK: GaloisField,
+    F: PrimeField,
+    C: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C, F> + Serializable + 'static,
 {
-    input_rows
-        .iter()
-        .map(|row| {
-            BitDecomposed::decompose(CV::BITS, |i| {
-                Replicated::new(
-                    Gf2::truncate_from(row.contribution_value.left()[i]),
-                    Gf2::truncate_from(row.contribution_value.right()[i]),
-                )
-            })
-        })
-        .collect::<Vec<_>>()
+    let check_times_value_ctx = ctx
+        .narrow(&Step::CheckTimesValue)
+        .set_total_records(num_records);
+
+    ctx.try_join(
+        check_bits
+            .into_iter()
+            .take(num_buckets)
+            .enumerate()
+            .map(|(check_idx, check)| {
+                let step = BitOpStep::from(check_idx);
+                let c = check_times_value_ctx.narrow(&step);
+                let record_id = RecordId::from(record_id);
+                let v = &value_bits;
+                async move {
+                    check
+                        .multiply(&v.to_additive_sharing_in_large_field(), c, record_id)
```

Comment:
We should not be doing this in a loop. We should do it only once outside of the loop.

Reply:
Funny.  I had the same comment, but I hadn't even thought of the performance implication, just that this function had no business dealing with the conversion.